openvdb/openvdb/tools/Morphology.h

```diff
@@ -63,540 +59,835 @@ enum NearestNeighbors { NN_FACE = 6, NN_FACE_EDGE = 18, NN_FACE_EDGE_VERTEX = 26
     using AccessorType = tree::ValueAccessor<TreeType>;
 
-    const bool   mOwnsManager;
-    ManagerType* mManager;
-    AccessorType mAcc;
-    int mSteps;
-
-    static const int LEAF_DIM     = LeafType::DIM;
-    static const int LEAF_LOG2DIM = LeafType::LOG2DIM;
-    using Word = typename DimToWord<LEAF_LOG2DIM>::Type;
-
-    struct Neighbor {
-        LeafType* leaf;//null if a tile
-        bool      init;//true if initialization is required
-        bool      isOn;//true if an active tile
-        Neighbor() : leaf(nullptr), init(true) {}
-        inline void clear() { leaf = nullptr; init = true; }
-        template<int DX, int DY, int DZ>
-        void scatter(AccessorType& acc, const Coord &xyz, int indx, Word mask)
+    Morphology(TreeType& tree)
+        : mManagerPtr(new tree::LeafManager<TreeType>(tree))
+        , mManager(*mManagerPtr)
+        , mThreaded(true) {}
+
+    Morphology(tree::LeafManager<TreeType>& tree)
+        : mManagerPtr(nullptr)
+        , mManager(tree)
+        , mThreaded(true) {}
+
+    /// @brief  Return whether this class is using multi-threading.
+    bool getThreaded() const { return mThreaded; }
+    /// @brief Set whether to use multi-threading.
+    /// @note  The grain size is not exposed
+    inline void setThreaded(const bool threaded) { mThreaded = threaded; }
+
+    /// @brief  Return a const reference to the leaf manager
+    inline const tree::LeafManager<TreeType>& manager() const { return mManager; }
+
+    /// @brief Topologically erode all voxels by the provided nearest neighbour
+    ///    scheme and optionally collapse constant leaf nodes
+    /// @details Inactive Tiles contribute to the erosion but active tiles are
+    ///    not modified.
+    /// @param iter Number of erosion iterations
+    /// @param nn Connectivity pattern of the erosion
+    /// @param prune Whether to collapse constant leaf nodes after the erosion
+    void erodeVoxels(const size_t iter,
+        const NearestNeighbors nn,
+        const bool prune = false);
+
+    /// @brief Topologically dilate all voxels by the provided nearest neighbour
+    ///    scheme and optionally collapse constant leaf nodes
+    /// @details Voxel values are unchanged and only leaf nodes are used to
+    ///    propagate the dilation.
+    /// @param iter Number of dilation iterations
+    /// @param nn Connectivity pattern of the dilation
+    /// @param prune Whether to collapse constant leaf nodes after the dilation
+    /// @param preserveMaskLeafNodes When dilating mask trees, the default behaviour
+    ///    chooses to steal the mask nodes rather than copy them. Although faster,
+    ///    this means that leaf nodes may be re-allocated. Set this to true if you
+    ///    need the original topology pointers to be preserved.
+    void dilateVoxels(const size_t iter,
+        const NearestNeighbors nn,
+        const bool prune = false,
+        const bool preserveMaskLeafNodes = false);
```

Comment:
Does this `preserveMaskLeafNodes` value actually get used anywhere? I can't seem to find any references to it being true.

Reply:
No it doesn't. This was introduced due a problem with the underlying implementation. This algorithm uses a mask topology  to perform the dilation. If the input is already a mask topology it just uses it directly. When dilating the mask topology, it steals the nodes and splits them across active threads. When it comes to re-combine the nodes, it's possible it'll favour newly allocated data over the input nodes, resulting in any held leaf pointers being invalidated. Users can use this internal option to retain the state for input mask dilation. It's impossible to know if this is required or not as the alternative is to deep copy the mask tree, a significant overhead. It's good to have this as an option (I think a problem with various bits of the VDB API is that it's sometimes clear what guarantees are provided from various functions w.r.t the state of the tree).

Reply:
I don't mind leaving this flag in, but I think if we include it without ever actually using it ourselves, it should at least be evaluated in the unit tests, which I don't think it is currently?

Reply:
Added a test